staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package environment
+
+import (
+	"fmt"
+
+	"github.com/google/cel-go/cel"
+)
+
+// Mode defines the supported environment modes for CEL in Kubernetes.
+type Mode string
+
+const (
+	// Supported provides a CEL environment mode configured with CEL language
+	// settings and libraries declared as supported by the current Kubernetes version.
+	//
+	// This environment must be used to validate (parse, compile, type check) all
+	// new or modified CEL expressions before they are written to storage. Already
+	// stored CEL expressions must not be validated with this environment since they
+	// have been written to storage by a newer version of Kubernetes. This
+	// environment must be compatible with the Supported environment of all previous
+	// versions of Kubernetes. All changes to this environment must first be
+	// included in Storage environment for at least one minor release before being
+	// included in this environment.
+	Supported Mode = "Supported"
+
+	// Storage provides a base CEL environment mode configured with CEL
+	// language settings and libraries compatible with all CEL expressions persisted
+	// to storage by all previous versions, the current version, and the next minor
+	// version of kubernetes. I.e. if 1.N is the current minor version, this
+	// environment supports versions 1.1...1.2, ... 1.(N-1), 1.N, 1.(N+1).
+	//
+	// To ensure safe rollback, this environment must be used to validate (parse,
+	// compile, type check), or evaluate CEL expressions from storage, since the
+	// expression might have been written to storage by a newer version of
+	// Kubernetes.
+	Storage Mode = "Storage"
+)
+
+// EnvSet provides a Kubernetes CEL environment set containing both a
+// "supported" and "storage" modes to enable controlled rollout of new
+// language settings and library functions.
+type EnvSet struct {
```

Comment:
if we keep track of the major/minor this envset should be compatible with, then Extend / ExtendProgram can take the major/minor along with the option and decide whether to add the given option to `Supported` or not

then consumers could use the same method to compatibly add things to their envs:

```
admissionEnv.Extend({1,28}, ... some option...)
admissionEnv.Extend({1,32}, ... some option...)
```

Reply:
This fell out nicely.